File: Day 19 07-11-24/solution-2.js
/*
Problem: First Unique Character in a String

Given a string s, find the first non-repeating character in it and return its index. If it does not exist, return -1.

Constraints:
1 <= s.length <= 10^5
s consists of only lowercase English letters.
*/

const firstUniqChar = (s) => {
    const charCount = {};

    // Count the occurrences of each character
    for (let char of s) {
        charCount[char] = (charCount[char] || 0) + 1;
    }

    // Find the first character with a count of 1
    for (let i = 0; i < s.length; i++) {
        if (charCount[s[i]] === 1) {
            return i;
        }
    }

    return -1;
};

File: Day 19 07-11-24/solution-2.py
class Solution:
    def firstUniqChar(self, s: str) -> int:
        """
        Find the first non-repeating character in a string and return its index.
        If it does not exist, return -1.

        Parameters:
        s (str): The input string.

        Returns:
        int: The index of the first non-repeating character or -1 if none exists.
        """
        # Dictionary to store the count of each character
        char_count = {}

        # Count the occurrences of each character
        for char in s:
            char_count[char] = char_count.get(char, 0) + 1

        # Find the first character with a count of 1
        for i, char in enumerate(s):
            if char_count[char] == 1:
                return i

        return -1
